tests/integration/istioio/trafficmanagement/scripts/traffic_shifting.txt

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 Istio Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -u
+set -o pipefail
+
+# setup bookinfo & sleep pods
+kubectl label namespace default istio-injection=enabled --overwrite || true
+
+kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml
+kubectl apply -f samples/bookinfo/networking/bookinfo-gateway.yaml
+kubectl apply -f samples/bookinfo/networking/destination-rule-all.yaml
+kubectl apply -f samples/sleep/sleep.yaml
+
+for deploy in "productpage-v1" "details-v1" "ratings-v1" "reviews-v1" "reviews-v2" "reviews-v3" "sleep"; do
+  if ! kubectl rollout status deployment "$deploy" --timeout 5m; then
+    echo "$deploy deployment rollout status check failed"
+    exit 1
+  fi
+done
+
+# Verification utils
+
+SLEEP_POD=$(kubectl get pod -l app=sleep -n default -o 'jsonpath={.items..metadata.name}')
+INGRESS_URL="http://istio-ingressgateway.istio-system"
+# reviews_v3_traffic_percentage
+# gets the % of productpage requests with reviews from reviews:v3 service
+function reviews_v3_traffic_percentage() {
+  set +e
+  local total_request_count=100
+  local v3_count=0
+  local v3_search_string="glyphicon glyphicon-star" # search string present in reviews_v3 response html
+  for ((i = 1; i <= total_request_count; i++)); do
+    if (kubectl exec "${SLEEP_POD}" -c sleep -n "default" -- curl -s $INGRESS_URL/productpage | grep -q "$v3_search_string"); then
+      v3_count=$((v3_count + 1))
+    fi
+  done
+  function is_in_range() {
+    local tol=10 #tolerance
+    local lower_bound=$(($2 - tol))
+    local upper_bound=$(($2 + tol))
+    if ((lower_bound < $1 && $1 < upper_bound)); then
+      return 0
+    fi
+    return 1
+  }
+  declare -a ranges=(0 25 50 75 100)
+  for i in "${ranges[@]}"; do
```

Comment:
@nmittler is this how we intend tests to be written? with all of the logic in bash?

Reply:
Right now, this is probably the best way to go. We likely need to extend the istio.io framework to automate some common tasks.

Reply:
This doesn't seem like a great design. Seems like we just threw out all the benefits of the test framework and are going in the exact opposite direction with writing code in bash, with antipatterns that have already been solved in the test framework like `sleep 10`. I thought the intent was to shell out only where the docs are running commands?

Reply:
@howardjohn the purpose of these tests is to verify the bash commands that we document on istio.io.  With that said, there will be boilerplate things we'll want to do like deploy things and wait until they're ready before we proceed.  AFAIK the options are:

Option | Limitations
--- | ---
Wait in code | This will have to be done before/after a script is run. So if a deploy/wait, for example, is done in the middle of a script, the test will have to split its script into 2 with a wait in between. This can undesirable/inflexible especially if there are many such cases in a single test.
Wait in script | This allows the entirety of a test to reside in a single bash script, if desired.  The disadvantage is that we're not leveraging many of the existing test utilities that exist in golang.

Ideally the script would just call out to the go code for common tasks. But this would either require a gRPC API or running binaries from the command line. Another alternative would be to rewrite some of those utilities in bash and inject the bash code when the commands are used. 

All of these options are a bit ugly. You have any thoughts or alternatives?

Reply:
I am trying to think of similar softwares, but I think all of those are single language (eg python Jupyter or https://scalameta.org/mdoc/).

I am not sure what is so bad about have the scripts split? I could image code like

```go
doSomeGoStuff()
runSnippet(`kubectl apply -f my-config/`, ValidationFunction)
goWait()
runSnippet(`curl foo`, ValidateCurl)
```

Does that make sense? I admittedly haven't looked too closely at how this is all implemented so maybe it doesn't make sense.

Reply:
@howardjohn technically, yes this would be a better solution if the test author is a SWE. The goal of this framework, however, is to allow non-SWEs (e.g. docs folks) to write these tests. In that case, they're not coders and golang may be a barrier of entry for them. The goal of the current design is to allow them to basically copy & paste the golang code and then just focus their effort on writing a bash script.

Now realistically, the first pass of the istio.io tests for our existing docs will likely be done by SWEs (since it's in our interest to automate this for releases). In the future, however, the hope is that the docs team will be able to write these tests as they author new content.

Reply:
Is any of the content we are testing really being written by strictly docs people? I assumed its generally SWEs implement a feature and write docs+tests for it. Is writing golang more of a barrier to entry than writing 100 lines of bash? To be honest, writing complex code in bash is a barrier of entry to me, not sure about others

Reply:
Are you asking who write the documentation on istio.io? I'm not sure who originally authored much of the content we have today, but IIUC the docs team own new content.

@rcaballeromx can keep me honest here

> To be honest, writing complex code in bash is a barrier of entry to me, not sure about others

Agreed, we should somehow eliminate the need for complex boilerplate in bash. This is something we can (and should) improve in the framework.

Reply:
I would expect a page like https://istio.io/docs/tasks/traffic-management/fault-injection/, something perfect to test, would be written by a dev (with help of course - we are not good writers :slightly_smiling_face: ), whereas concept type docs are mostly just docs writers, but I could be way off here.

If we make common tasks simple and still use bash then its fine with me

Reply:
@nmittler @howardjohn we should really have this discussion in the doc WG, instead of in PR comments, but here's my 2 cents worth.

IMO, docs need to be written by somebody that is good at explaining things and not necessarily a strong developer. Some docs will be written by devs, some by doc writers, some by people that are both. Getting good doc published quickly is the primary goal, so writing the test is overhead for every writer, regardless of programming ability, so it needs to be minimized. I would say the ultimate goal, when or if we ever do it, is to be able to generate a test from an annotated markdown file. (As an aside, that's why I'm still convinced the code snippets should be generated from the docs, not the other way around). All the surrounding boilerplate (setup, wait, etc.) should either be generated from annotations, or really easy to invoke, whether that's go code or bash doesn't matter to me. Depending on the individual one will be better than the other, but both are a pain if they involve more than trivial invoke like calls.

Let's use https://istio.io/docs/tasks/traffic-management/fault-injection/ as an example, as @howardjohn suggested. Btw, I wrote the first version of this, and several other traffic management tasks and examples. Like most of the docs we want to test, it's a recipe of commands and checks. First step for the test ("Before you begin" section) is to start up the bookinfo app (1 line call to a function to start bookinfo - can be reused in lots of tests), and then invoke a couple of kubectl commands to configure the routing (call bash commands). Next ("Injecting an HTTP delay fault" section) we invoke another bash command, and a command to confirm it's set (interesting that the doc uses `...` elision in the expected output to make the doc easier to read, so would be good to support elided text comparison in the framework). Next we need to wait for a few seconds and then check for the error described in "Testing the delay configuration" section. This is the first part that needs some actual code to simulate the same conditions as described in the doc using a browser, although things like login as user jason, and time and check the response of accessing a page, would definitely be generic functions that could be used in several tests. So the only question is what is the list of primitive functions that are available and how are they invoked.

Making these invocations as easy as possible (generating the calls from annotations being the ultimate) is what I think we need if we're really going to have the majority of our docs be well tested, which would be fantastic.